narwhals/dependencies.py

```diff
@@ -15,6 +19,15 @@
     import pandas
 
 
+class Implementation(Enum):
+    POLARS = auto()
+    PANDAS = auto()
+    MODIN = auto()
+    CUDF = auto()
+    PYARROW = auto()
+    NUMPY = auto()
```

Comment:
what's this?

Reply:
This is the enum where we register the different implementations, whenever we add a new one, we need to add it here and MyPy will flag all the necessary changes

Reply:
Sorry I mean, why numpy?

Reply:
Because there are a couple of places where we call  `get_backend` with `Implementation.Numpy`. The idea behind the enums is to always access the compute backend through `get_backend`.

This is not fully done yet, but after this PR is merge, I can work on replacing all the intances of `get_*` with the `get_backend`. At the moment there is only two places, which are the entrypoints.

The rationale behind this is to make the code fully compute backend agnostic

Reply:
sure, but we don't have a numpy backend?

I mean, pandas uses it under-the-hood, but as far as narwhals is concerned, the dataframe backend is pandas, not numpy

Reply:
That is true, I just wanted to abstract the way we interact with dependencies from within the package to avoid calling `get_numpy` directly and use the `get_backend` function. This may be a little misleading as Numpy is not a backend pre se. It was just for convenience.

However to further limit the scope of what is and what is not a dataframe engine, I created two types below `DATAFRAME_IMPLEMENTATIONS` and `SERIES_IMPLEMENTATIONS` which do not include Numpy. They are both thought to be subset of this `Implementation` enum. That means though that maybe `Implementation` is not the best name for this.

I wanted the enum to capture everything that was external, be it an auxiliary library like numpy or pyarrow_compute or a dataframe engine like pandas

Reply:
I don't think numpy and pyarrow.compute really belong here

I'd suggest just having `Implementation` here, and then `PandasImplementation`

Reply:
And to make sure that both `pyarrow.compute` and `numpy` are loadable by the `get_backend` function, how should we name that Enum? It should be different than `Implementation`

Reply:
we can just use `get_numpy` / `get_pyarrow_compute`

Reply:
Done